extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -0,0 +1,359 @@
+
+    /** @private {?../../../src/service/action-impl.ActionService} */
+    this.action_ = null;
+
+    /** @private {number|string} */
+    this.prefix_ = element.id || Math.floor(Math.random() * 100);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout === Layout.FIXED_HEIGHT;
+  }
+
+  /** @override */
+  buildCallback() {
+    // TODO(#24898): remove this assert when cleaning up experiment post launch.
+    userAssert(
+      isExperimentOn(this.win, 'amp-mega-menu'),
+      'Turning on the amp-mega-menu experiment is necessary to use the amp-mega-menu component.'
+    );
+
+    this.action_ = Services.actionServiceForDoc(this.element);
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.registerMenuItems_();
+    // items may not be present after build if dynamically rendered via amp-list,
+    // in which case register them after DOM update instead.
+    this.element.addEventListener(AmpEvents.DOM_UPDATE, () => {
+      this.registerMenuItems_();
+    });
+
+    const mask = this.document_.createElement('div');
+    mask.classList.add('i-amphtml-mega-menu-mask');
+    mask.setAttribute('aria-hidden', 'true');
+    // append mask to header so that all children of header appear above the mask
+    const maskParent =
+      closestAncestorElementBySelector(this.element, 'header') || this.element;
+    maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
+    maskParent.appendChild(mask);
+    this.maskElement_ = mask;
+
+    this.documentElement_.addEventListener('click', () => this.collapse_());
+    this.documentElement_.addEventListener('keydown', event => {
+      // Collapse mega menu on ESC.
+      if (event.key === Keys.ESCAPE && this.collapse_()) {
+        event.preventDefault();
+      }
+    });
+
+    return Promise.resolve();
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    // TODO(#25047): unregister event listeners.
+    // Ensure that menu is closed when hidden via media query.
+    this.collapse_();
+    return false;
+  }
+
+  /**
+   * Find all expandable menu items under this mega menu and add appropriate
+   * classes, attributes and event listeners to its children.
+   * @private
+   */
+  registerMenuItems_() {
+    this.items_ = toArray(scopedQuerySelectorAll(this.element, 'nav > * > li'));
+    this.items_.forEach(item => {
+      // skip if the item has already been registered.
+      if (item.classList.contains('i-amphtml-mega-menu-item')) {
+        return;
+      }
+      const heading =
+        scopedQuerySelector(item, '> button') ||
+        scopedQuerySelector(item, '> [role=button]');
+      const content = scopedQuerySelector(item, '> [role=dialog]');
+      // do not register item if either its heading or content element is missing,
+      // or if heading has on tap action.
+      if (
+        !heading ||
+        !content ||
+        this.action_.hasAction(heading, 'tap', item)
+      ) {
+        return;
+      }
```

Comment:
Do we still want the item in `this.items_` if it's not really registered? If not, we can get rid of `this.itemCount_`